### PR TITLE
EOS-28494: Fix the MAX_KEY_SIZE issue for vkvv format

### PR DESCRIPTION
**Pull request by RadhaGulhane13**
Signed-off-by: Radha Gulhane <radha.gulhane@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: btree/btree.c [hunk 1]
```diff
@@ -1557,6 +1568,13 @@ static int bnode_valsize(const struct nd *node)
 	return (node->n_type->nt_valsize(node));
 }
 
+static int bnode_max_ksize(const struct nd *node)
+{
+	M0_PRE(bnode_invariant(node));
+	return (node->n_type->nt_max_ksize(node));
+}
+
```
Comment (ydb242) [inline btree/btree.c:1576]: Extra lines can be removed

File: btree/btree.c [hunk 2]
```diff
@@ -5023,6 +5066,8 @@ static void vkvv_lnode_make(struct slot *slot)
 		m0_memmove(start_val_addr - total_vsize - vsize,
 			   start_val_addr - total_vsize, total_vsize);
 	}
+	if (ksize > h->vkvv_max_ksize)
```
Comment (ydb242) [inline btree/btree.c:5069]: Add a line above this statement

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: btree/btree.c [hunk 3]
```diff
@@ -924,7 +929,7 @@ struct node_type {
 	bool  (*nt_isunderflow)(const struct nd *node, bool predict);
 
 	/** Returns true if there is possibility of overflow. */
-	bool  (*nt_isoverflow)(const struct nd *node,
+	bool  (*nt_isoverflow)(const struct nd *node, int max_ksize,
```
[the remark this answers is not in the document]
Comment (RadhaGulhane13) [reply btree/btree.c:938]: Updated comments

File: btree/btree.c (hunk 1 above)
Comment (RadhaGulhane13) [reply btree/btree.c:1582]: done

File: btree/btree.c (hunk 2 above)
Comment (RadhaGulhane13) [reply btree/btree.c:5080]: done

File: btree/btree.c [hunk 4]
```diff
@@ -928,7 +928,13 @@ struct node_type {
 	 */
 	bool  (*nt_isunderflow)(const struct nd *node, bool predict);
 
-	/** Returns true if there is possibility of overflow. */
+	/**
+	 * Returns true if there is possibility of overflow.
+	 * The parameter max_ksize will be used only in the case of internal
+	 * nodes of variable sized keys and values node format. In worst case
+	 * scenario, size of the key which is required to get added in internal
```
Comment (shashank-parulekar) [inline btree/btree.c:935]: s/to get added/to be added/

File: btree/btree.c [hunk 5]
```diff
@@ -928,7 +928,13 @@ struct node_type {
 	 */
 	bool  (*nt_isunderflow)(const struct nd *node, bool predict);
 
-	/** Returns true if there is possibility of overflow. */
+	/**
+	 * Returns true if there is possibility of overflow.
+	 * The parameter max_ksize will be used only in the case of internal
+	 * nodes of variable sized keys and values node format. In worst case
```
Comment (shashank-parulekar) [inline btree/btree.c:934]: s/In worst case scenario//

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: btree/btree.c (hunk 4 above)
Comment (RadhaGulhane13) [reply btree/btree.c]: done

File: btree/btree.c (hunk 5 above)
Comment (RadhaGulhane13) [reply btree/btree.c]: done

**Review by shashank-parulekar — APPROVED** [review]

File: btree/btree.c [hunk 6]
```diff
@@ -2679,7 +2714,7 @@ static bool ff_isunderflow(const struct nd *node, bool predict)
 	return  rec_count == 0;
 }
 
-static bool ff_isoverflow(const struct nd *node, const struct m0_btree_rec *rec)
+static bool ff_isoverflow(const struct nd *node, int max_ksize, const struct m0_btree_rec *rec)
```
Comment (ydb242) [inline btree/btree.c:2717]: 80 char limit

**Review by ydb242 — APPROVED** [review]

**Pull request merged**